tutorials/baxus.ipynb

```diff
@@ -0,0 +1,985 @@
+    "            -1\n",
+    "            * (self.new_bins_on_split * self.eval_budget * self.target_dim)\n",
+    "            / (self.d_init * (1 - (self.new_bins_on_split + 1) ** (self.n_splits + 1)))\n",
+    "        )\n",
+    "\n",
+    "    @property\n",
+    "    def failure_tolerance(self) -> int:\n",
+    "        if self.target_dim == self.dim:\n",
+    "            return self.target_dim\n",
+    "        k = math.floor(math.log(self.length_min / self.length_init, 0.5))\n",
+    "        split_budget = self.split_budget\n",
+    "        return min(self.target_dim, max(1, math.floor(split_budget / k)))\n",
+    "\n",
+    "\n",
+    "def update_state(state, Y_next):\n",
+    "    if max(Y_next) > state.best_value + 1e-3 * math.fabs(state.best_value):\n",
+    "        state.success_counter += 1\n",
+    "        state.failure_counter = 0\n",
+    "    else:\n",
+    "        state.success_counter = 0\n",
+    "        state.failure_counter += 1\n",
+    "\n",
+    "    if state.success_counter == state.success_tolerance:  # Expand trust region\n",
+    "        state.length = min(2.0 * state.length, state.length_max)\n",
+    "        state.success_counter = 0\n",
+    "    elif state.failure_counter == state.failure_tolerance:  # Shrink trust region\n",
+    "        state.length /= 2.0\n",
+    "        state.failure_counter = 0\n",
+    "\n",
+    "    state.best_value = max(state.best_value, max(Y_next).item())\n",
+    "    if state.length < state.length_min:\n",
+    "        state.restart_triggered = True\n",
+    "    return state"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Create a BAxUS embedding\n",
+    "\n",
+    "We now show how to create the BAxUS embedding. The essential idea is to assign input dimensions to target dimensions and to assign a sign $\\in \\pm 1$ to each input dimension, similar to the HeSBO embedding. \n",
+    "We create the embedding matrix that is used to project points from the target to the input space. The matrix is sparse, each column has precisely one non-zero entry that is either 1 or -1."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "tensor([[ 0.,  0.,  0.,  0.,  0.,  1.,  1.,  0.,  1., -1.],\n",
+       "        [ 0.,  1., -1.,  0., -1.,  0.,  0.,  0.,  0.,  0.],\n",
+       "        [-1.,  0.,  0., -1.,  0.,  0.,  0., -1.,  0.,  0.]], device='cuda:0',\n",
+       "       dtype=torch.float64)"
+      ]
+     },
+     "execution_count": 5,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "def embedding_matrix(input_dim: int, target_dim: int) -> torch.Tensor:\n",
+    "\n",
+    "    if (\n",
+    "        target_dim >= input_dim\n",
+    "    ):  # return identity matrix if target size greater than input size\n",
+    "        return torch.eye(input_dim).to(device=device, dtype=dtype)\n",
```

Comment:
There are a number of other places in this PR where a `.to` isn't necessary.

```suggestion
    "        return torch.eye(input_dim, device=device, dtype=dtype)\n",
```

Reply:
Thanks, I think I fixed this where possible